wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/agent/server/WsAgentHealthCheckerImpl.java

```diff
@@ -0,0 +1,118 @@
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.UnauthorizedException;
+import org.eclipse.che.api.core.model.machine.Machine;
+import org.eclipse.che.api.core.model.machine.Server;
+import org.eclipse.che.api.core.rest.HttpJsonRequest;
+import org.eclipse.che.api.core.rest.HttpJsonRequestFactory;
+import org.eclipse.che.api.core.rest.HttpJsonResponse;
+import org.eclipse.che.api.machine.shared.Constants;
+import org.eclipse.che.api.workspace.shared.dto.WsAgentHealthStateDto;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import javax.ws.rs.HttpMethod;
+import java.io.IOException;
+import java.util.Map;
+
+import static javax.ws.rs.core.Response.Status.NOT_FOUND;
+import static javax.ws.rs.core.Response.Status.SERVICE_UNAVAILABLE;
+import static org.eclipse.che.api.machine.shared.Constants.WSAGENT_REFERENCE;
+import static org.eclipse.che.dto.server.DtoFactory.newDto;
+
+/**
+ * Mechanism for checking workspace agent's state.
+ *
+ * @author Vitalii Parfonov
+ * @author Valeriy Svydenko
+ */
+@Singleton
+public class WsAgentHealthCheckerImpl implements WsAgentHealthChecker {
+    private static final String WS_AGENT_SERVER_NOT_FOUND_ERROR = "Workspace agent server not found in dev machine.";
+
+    protected static final Logger LOG = LoggerFactory.getLogger(WsAgentHealthCheckerImpl.class);
+
+    private final HttpJsonRequestFactory httpJsonRequestFactory;
+    private final int                    wsAgentPingConnectionTimeoutMs;
+
+    @Inject
+    public WsAgentHealthCheckerImpl(HttpJsonRequestFactory httpJsonRequestFactory,
+                                    @Named("machine.ws_agent.ping_conn_timeout_ms") int wsAgentPingConnectionTimeoutMs) {
+        this.httpJsonRequestFactory = httpJsonRequestFactory;
+        this.wsAgentPingConnectionTimeoutMs = wsAgentPingConnectionTimeoutMs;
+    }
+
+    @Override
+    public WsAgentHealthStateDto check(Machine machine) throws NotFoundException,
+                                                               ServerException,
+                                                               ForbiddenException,
+                                                               BadRequestException,
+                                                               UnauthorizedException,
+                                                               ConflictException {
+        final Map<String, ? extends Server> servers = machine.getRuntime().getServers();
+        Server wsAgent = getWsAgent(servers);
+        final WsAgentHealthStateDto agentHealthStateDto = newDto(WsAgentHealthStateDto.class);
+        if (wsAgent == null) {
+            return agentHealthStateDto.withCode(NOT_FOUND.getStatusCode())
+                                      .withReason("Workspace Agent not available if Dev machine are not RUNNING");
```

Comment:
It is not obvious why this code makes assumption that provided machine is dev-machine and that it is not in RUNNING state